worker/peergrouper/desired.go

```diff
@@ -411,29 +390,95 @@ func (p *peerGroupChanges) createNonVotingMember(
-			continue
+// updateAddresses updates the member addresses in the new replica-set, using
+// the HA space if one is configured.
+func (p *peerGroupChanges) updateAddresses(info *peerGroupInfo) error {
+	update := p.updateAddressFromSpace
+	if info.haSpace == "" {
+		update = p.updateAddressFromInternal
+	}
+
+	for id := range p.members {
+		if err := update(id, info); err != nil {
+			return errors.Annotate(err, "updating member address")
 		}
-		if addr != p.members[id].Address {
-			p.members[id].Address = addr
+	}
+	return nil
+}
+
+// updateAddressUsingSpace sets the member address based on the configured
+// HA space. If no addresses are available for the machine and space,
+// then an error is returned.
+func (p *peerGroupChanges) updateAddressFromSpace(id string, info *peerGroupInfo) error {
+	m := info.machines[id]
+	addr, err := m.SelectMongoAddressFromSpace(info.mongoPort, info.haSpace)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if addr != p.members[id].Address {
+		p.members[id].Address = addr
+		p.isChanged = true
+	}
+	return nil
+}
+
+// updateAddressFromInternal attempts to update the member with a cloud-local
+// address from the machine.
+// If there is a single cloud local address available, it is used.
+// If there are multiple addresses, then a check is made to ensure that:
+//   - the member was previously in the replica-set and;
+//   - the previous address used for replication is still available.
+// If the check is satisfied, then a warning is logged and no change is made.
+// Otherwise an error is returned to indicate that a HA space must be
+// configured in order to proceed.
+func (p *peerGroupChanges) updateAddressFromInternal(id string, info *peerGroupInfo) error {
+	member := p.members[id]
+	hostPorts := info.machines[id].GetPotentialMongoHostPorts(info.mongoPort)
+	addrs := network.SelectInternalHostPorts(hostPorts, false)
+
+	// This should not happen because SelectInternalHostPorts will choose a
+	// public address when there are no cloud-local addresses.
+	// Zero addresses would mean the machine would not be accessible at all.
+	// We ignore this outcome and leave the address alone.
+	if len(addrs) == 0 {
+		return nil
+	}
+
+	if len(addrs) == 1 {
+		addr := addrs[0]
+		logger.Debugf("machine %q selected address %q by scope from %v", id, addr, hostPorts)
+
+		if member.Address != addr {
+			member.Address = addr
 			p.isChanged = true
 		}
+		return nil
+	}
+
+	msg := fmt.Sprintf("machine %q has more than one non-local address and juju-ha-space is not set", id)
+
+	// If this would-be member is new, we enforce the policy of having a
+	// configured HA space when there are multiple cloud-local addresses.
+	if _, ok := info.recognised[id]; !ok {
+		return errors.New(msg)
 	}
+
+	// If the prior address is still available, we allow preservation of the
+	// status quo rather than throwing out with an error.
+	// TODO (manadart 2018-03-21) Should we also check the voting status?
```

Comment:
Given the general goal of appearing to be a valid host, I don't think we have to only do this when the controller is already voting.
A warning either way allows you to do something about it, without the peergrouper dying, preventing you from making progress.